### PR TITLE
Apg/sap integration

### PR DESCRIPTION
**Pull request by annapiegra**
## PR Summary
Creates a new f90-interface needed for the Pagosa integration.

-Adds eos_setOptions 
-Decouples eospac preinversion and insert data
-Uses a single step instead of two to get sesame P from D+Ut 

TODO: docs, tests, fix properly singularity_eos.f90 spiner deps...

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file
- [ ] If preparing for a new release, update the version in cmake.

### CODE REVIEW

**Comment by jhp-lanl** [comment]
@annapiegra I fixed a couple errors that were breaking the build. I also added the `MinInternalEnergyFromDensity()` function to the scratch buffer size querrying mechanism. Currently I'm assuming EOSPAC either won't write to the second derivative array (`dFy` in EOSPAC parlance) or that because we don't care about that value, nothing nasty will happen when it does. Feel free to add another scratch buffer if you think that's going to be an issue.

**Comment by annapiegra** [comment]
Thanks.  Since it’s a 1d function dFy shouldn’t matter. Also we don’t use the derivatives so we should be good on that front too.I’ll verify later this week. What compiler were you using? I’ve only tested with the intel compiler used by sap.

Sent with BlackBerry Work
(www.blackberry.com)


From: Jeff Peterson ***@***.******@***.***>>
Date: Tuesday, Jul 25, 2023 at 8:41 PM
To: lanl/singularity-eos ***@***.******@***.***>>
Cc: Pietarila Graham, Anna Mataleena ***@***.******@***.***>>, Mention ***@***.******@***.***>>
Subject: [EXTERNAL] Re: [lanl/singularity-eos] Apg/sap integration (PR #278)


@annapiegra<https://urldefense.com/v3/__https://github.com/annapiegra__;!!Bt8fGhp8LhKGRg!FyhuaD2hC3HpePc4nPXZnp4_hKXTt7A0ArtoSCtcAFKj4jw-LI02krJqVh9TevMmKv24Pi7JVIRigWte8djEk7o$> I fixed a couple errors that were breaking the build. I also added the MinInternalEnergyFromDensity() function to the scratch buffer size querrying mechanism. Currently I'm assuming EOSPAC either won't write to the second derivative array (dFy in EOSPAC parlance) or that because we don't care about that value, nothing nasty will happen when it does. Feel free to add another scratch buffer if you think that's going to be an issue.

File: eospac-wrapper/eospac_wrapper.cpp [hunk 1]
```diff
@@ -127,19 +129,78 @@ EOS_INTEGER eosSafeLoad(int ntables, int matid, EOS_INTEGER tableType[],
...
+    EOS_REAL values[] = {insert_data};
+    for (int i = 0; i < ntables; i++) {
+      eos_SetOption(&(tableHandle[i]), &EOS_INSERT_DATA, &(values[0]), &errorCode);
+      eosCheckError(errorCode, "eospac options: eos_insert_data", eospacWarn);
+    } 
+  }
+
+
+  //choice of which table types setOption is called on mimics SAP. Some table types are incmopatible whiel others lead to numerical issues. 
+  if (monotonicity==1 || monotonicity==2) {
```
Comment (Yurlungur) [inline eospac-wrapper/eospac_wrapper.cpp:151]: What is the meaning of these settings? It might be better if we used an enum for monotonicity so we could more easily read what these settings do?

File: eospac-wrapper/eospac_wrapper.cpp [hunk 2]
```diff
@@ -127,19 +129,78 @@ EOS_INTEGER eosSafeLoad(int ntables, int matid, EOS_INTEGER tableType[],
...
+  if (apply_smoothing) {
+    for (int i = 0; i < ntables; i++) {
+      if (tableType[i]==EOS_Pt_DUt || tableType[i]==EOS_T_DUt || tableType[i]==EOS_Ut_DPt){
+	eos_SetOption(&(tableHandle[i]), &EOS_SMOOTH,  &(EOS_NullVal), &errorCode);
+	eosCheckError(errorCode, "eospac options: eos_smooth", eospacWarn);
+      }
+    }
+  }
+ 
+  if (abs(apply_splitting) == 1) {
```
Comment (Yurlungur) [inline eospac-wrapper/eospac_wrapper.cpp:176]: same as above---maybe an enum would be better here?

**Review by Yurlungur — APPROVED** [review]
Overall this seems reasonable to me. I'd like Richard's eyes on the EOSPAC stuff and I would also like to make sure all the PR checklist boxes are checked. But barring that, I approve. A few nitpicks below.

File: eospac-wrapper/eospac_wrapper.cpp (hunk 1 above)
Comment (annapiegra) [reply eospac-wrapper/eospac_wrapper.cpp]: Added enum

File: eospac-wrapper/eospac_wrapper.cpp (hunk 2 above)
Comment (annapiegra) [reply eospac-wrapper/eospac_wrapper.cpp]: added enum

File: singularity-eos/eos/eos_gruneisen.hpp [hunk 3]
```diff
@@ -313,6 +314,11 @@ PORTABLE_INLINE_FUNCTION Real Gruneisen::PressureFromDensityInternalEnergy(
   }
   return P_H + Gamma(rho) * rho * (sie - E_H);
 }
+PORTABLE_INLINE_FUNCTION Real Gruneisen::MinInternalEnergyFromDensity(
+    const Real rho_in, Real *lambda) const {
+  const Real rho = std::min(rho_in, _rho_max);
```
[the remark this answers is not in the document]
Comment (annapiegra) [reply singularity-eos/eos/eos_gruneisen.hpp:320]: Added.

File: singularity-eos/eos/eos_base.hpp [hunk 4]
```diff
@@ -262,6 +263,35 @@ class EosBase {
...
+                                 std::forward<LambdaIndexer>(lambdas));
+  }
+  template <typename LambdaIndexer>
+  inline void MinInternalEnergyFromDensity(const Real *rhos, Real *sies,
+                                           Real * /*scratch*/, const int num,
+                                           LambdaIndexer &&lambdas,
+                                           Transform && = Transform()) const {
+    MinInternalEnergyFromDensity(rhos, num, std::forward<LambdaIndexer>(lambdas));
+  }
+  ///
```
Comment (jhp-lanl) [inline singularity-eos/eos/eos_base.hpp:294]: Would it make sense to make a default definition of `MinInternalEnergyFromDensity()` in the EosBase class? It could be something as simple as taking the zero-K isotherm energy at a given density. EOSPAC would of course override this, but I think it would be effective for many of the analytic EOS rather than just returning zero.

Comment (annapiegra) [reply singularity-eos/eos/eos_base.hpp:294]: Is this something we want implemented asap?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:294]: I guess since Pagosa only does this for sesame tables, it's not a very big deal right now. But I would _greatly_ prefer that approach to stubbing out the feature for all EOS. It should only be a few lines of code that need to be changed though.

**Comment by annapiegra** [comment]
Formatting check fails because changing the order of hpp files in eospac_wrapper.cpp leads to compiler errors.

**Comment by jhp-lanl** [comment]
@cmauney @Yurlungur @dholladay00 @rbberger  I think this is ready to go if any of you have further comments

File: singularity-eos/eos/eos_base.hpp (hunk 4 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:304]: Made an issue for this https://github.com/lanl/singularity-eos/issues/318

File: singularity-eos/eos/modifiers/ramps_eos.hpp [hunk 5]
```diff
@@ -117,7 +117,7 @@ class BilinearRampEOS : public EosBase<BilinearRampEOS<T>> {
 
   PORTABLE_INLINE_FUNCTION
   Real get_ramp_dpdrho(Real rho) const {
-    const Real dpdr{rho < r0_ ? 0.0 : rho < rmid_ ? ratio(a_, r0_) : ratio(b_, r0_)};
+    const Real dpdr{rho < rmid_ ? ratio(a_, r0_) : ratio(b_, r0_)};
```
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:120]: Consider removing the branching here if possible by using masks instead?

**Review by Yurlungur — APPROVED** [review]
Great work all---glad to see this in such good shape. I will let @mauneyc-LANL and @dholladay00 weigh in, but from my perspective, this is ready to go.

File: singularity-eos/eos/modifiers/ramps_eos.hpp (hunk 5 above)
Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:120]: I think we'll defer this since it might make sense to replace these all wholesale in the future.

Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:120]: see https://github.com/lanl/singularity-eos/issues/319

**Comment by jhp-lanl** [comment]
@cmauney @Yurlungur @dholladay00 @rbberger  can I interpret your silence for approval?

File: singularity-eos/eos/eos_ideal.hpp [hunk 6]
```diff
@@ -78,6 +78,11 @@ class IdealGas : public EosBase<IdealGas> {
       const Real rho, const Real sie, Real *lambda = nullptr) const {
     return MYMAX(0.0, _gm1 * rho * sie);
   }
+  PORTABLE_INLINE_FUNCTION Real
+  MinInternalEnergyFromDensity(const Real rho, Real *lambda = nullptr) const {
+    return 0.0;
```
Comment (dholladay00) [inline singularity-eos/eos/eos_ideal.hpp:83]: @jhp-lanl could you add the not enabled warning function here?

File: singularity-eos/eos/eos_vinet.hpp [hunk 7]
```diff
@@ -304,6 +306,14 @@ PORTABLE_INLINE_FUNCTION Real Vinet::PressureFromDensityInternalEnergy(
   Vinet_F_DT_func(rho, temp, output);
   return output[1];
 }
+PORTABLE_INLINE_FUNCTION Real Vinet::MinInternalEnergyFromDensity(const Real rho,
+                                                                  Real *lambda) const {
+#ifndef NDEBUG
```
Comment (dholladay00) [inline singularity-eos/eos/eos_vinet.hpp:311]: could probably replace with the function @jhp-lanl

File: singularity-eos/eos/eos_ideal.hpp (hunk 6 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_ideal.hpp:83]: This value is actually valid for the ideal gas. However, an appropriate shift was not implemented in the Shift modifier (which is the only thing that would cause this value to be non-zero). I've added that now

File: singularity-eos/eos/eos_vinet.hpp (hunk 7 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_vinet.hpp]: done

File: spack-repo/packages/singularity-eos/package.py [hunk 8]
```diff
@@ -12,9 +12,9 @@ class SingularityEos(CMakePackage, CudaPackage):
     """Singularity-EOS: A collection of closure models and tools useful for
     multiphysics codes."""
 
-    homepage = "https://lanl.github.io/singularity-eos/main/index.html"
-    git = "https://github.com/lanl/singularity-eos.git"
-    url = "https://github.com/lanl/singularity-eos/archive/refs/tags/release-1.6.1.tar.gz"
+    homepage = "https://re-git.lanl.gov/xcap/oss/singularity-eos"
```
Comment (rbberger) [inline spack-repo/packages/singularity-eos/package.py:15]: I don't think these 3 lines should change

File: singularity-eos/eos/eos_eospac.hpp [hunk 9]
```diff
@@ -53,10 +54,91 @@ using namespace eos_base;
 #define SG_PIF_NOWARN
 #endif // !defind SG_PIF_NOWARN
 
+namespace impl_eospac {
+
+inline void SetUpDensityTemperatureScalingOptions(EOS_INTEGER options[],
+                                                  EOS_REAL values[], EOS_INTEGER &nopts,
+                                                  Transform &transform) {
```
Comment (rbberger) [inline singularity-eos/eos/eos_eospac.hpp:61]: This and others can be `const Transform &`

File: singularity-eos/eos/eos_eospac.hpp [hunk 10]
```diff
@@ -590,72 +638,58 @@ class EOSPAC : public EosBase<EOSPAC> {
...
-      EOS_REAL values[2];
-      EOS_INTEGER nopts = 0;
+  template <typename LambdaIndexer>
+  inline void MinInternalEnergyFromDensity(const Real *rhos, Real *sies, Real *scratch,
+                                           const int num, LambdaIndexer /*lambdas*/,
+                                           Transform &&transform = Transform()) const {
+    using namespace EospacWrapper;
+    EOS_REAL *R = const_cast<EOS_REAL *>(&rhos[0]);
+    EOS_REAL *E = &sies[0];
+    EOS_REAL *dedr = scratch + 1 * num;
```
Comment (rbberger) [inline singularity-eos/eos/eos_eospac.hpp:667]:
```suggestion
    EOS_REAL *dedr = scratch + 0 * num;
```

**Review by rbberger — APPROVED** [review]
Overall looks good to me. Just some minor things I found.

File: singularity-eos/eos/eos_base.hpp [hunk 11]
```diff
@@ -68,6 +68,7 @@ char *StrCat(char *destination, const char *source) {
   using EosBase<EOSDERIVED>::InternalEnergyFromDensityTemperature;                       \
   using EosBase<EOSDERIVED>::PressureFromDensityTemperature;                             \
   using EosBase<EOSDERIVED>::PressureFromDensityInternalEnergy;                          \
+  using EosBase<EOSDERIVED>::MinInternalEnergyFromDensity;                               \
```
Comment (rbberger) [inline singularity-eos/eos/eos_base.hpp:71]: Still needs to be added to the Python interface, but that can be its own PR.

File: singularity-eos/eos/eos_eospac.hpp [hunk 12]
```diff
@@ -1331,8 +1323,31 @@ SG_PIF_NOWARN
 PORTABLE_INLINE_FUNCTION Real EOSPAC::PressureFromDensityInternalEnergy(
     const Real rho, const Real sie, Real *lambda) const {
   using namespace EospacWrapper;
-  Real temp = TemperatureFromDensityInternalEnergy(rho, sie, lambda);
-  return PressureFromDensityTemperature(rho, temp, lambda);
+  EOS_INTEGER options[]{EOS_Y_CONVERT, EOS_F_CONVERT};
+  EOS_REAL values[]{sieFromSesame(1.0), pressureFromSesame(1.0)};
+  EOS_INTEGER nopts = 2;
+  EOS_REAL R[1] = {rho}, E[1] = {sie}, P[1], dPdr[1], dPde[1], Ec[1];
```
Comment (dholladay00) [inline singularity-eos/eos/eos_eospac.hpp:1329]: I'm seeing an unused variable warning on `Ec`

File: spack-repo/packages/singularity-eos/package.py (hunk 8 above)
Comment (jhp-lanl) [reply spack-repo/packages/singularity-eos/package.py]: Fixed

File: singularity-eos/eos/eos_eospac.hpp (hunk 9 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp]: Done

File: singularity-eos/eos/eos_base.hpp (hunk 11 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:71]: Issue here: https://github.com/lanl/singularity-eos/issues/321

File: singularity-eos/eos/eos_eospac.hpp (hunk 12 above)
Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp:1330]: I'm surprised I didn't get that. What did you need to do to enable that warning?

The unused variable is fixed though

Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp]: Along these lines, should we enable `-Werror` in the CI?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_eospac.hpp]: @dholladay00 I'm going to go ahead and merge. I fixed the unused variable, but we should discuss whether we should maybe change the CMake to error on warnings and make sure that all appropriate warnings are enabled.

**Comment by jhp-lanl** [comment]
Tests pass on gitlab as well as github. Merging.
